[Remote Inspection] Make it possible to target ::before/::after pseudo elements

whsieh opened the pull request:
#### 780bd128913de133fa2bcadf4fdd2127be890bcb
<pre>
[Remote Inspection] Make it possible to target ::before/::after pseudo elements
<a href="https://bugs.webkit.org/show_bug.cgi?id=271791">https://bugs.webkit.org/show_bug.cgi?id=271791</a>

Reviewed by Megan Gardner.

(...and also incorporates suggestions from Antti).

Add support for targeting `::before` and `::after` pseudo elements, for visibility adjustment. See
below for more details.

* LayoutTests/fast/element-targeting/target-pseudo-elements-expected.html: Added.
* LayoutTests/fast/element-targeting/target-pseudo-elements.html: Added.

Add a new layout test to exercise the change (targeting both `after` and `before` pseudo elements).

* Source/WebCore/Headers.cmake:

Add a new WebCore header that just contains the new `VisibilityAdjustment` enum values.

* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/dom/Element.cpp:
(WebCore::Element::visibilityAdjustment const):
(WebCore::Element::addVisibilityAdjustment):
(WebCore::Element::isVisibilityAdjustmentRoot const): Deleted.
(WebCore::Element::setIsVisibilityAdjustmentRoot): Deleted.

Refactor this code — instead of having a single boolean flag that indicates whether the element is
a visibility adjustment root, turn it into an OptionSet with up to 3 bits (subtree, `::before`
pseudo element and `::after` pseudo element).

* Source/WebCore/dom/Element.h:
* Source/WebCore/dom/ElementRareData.cpp:
* Source/WebCore/dom/ElementRareData.h:
(WebCore::ElementRareData::visibilityAdjustment const):
(WebCore::ElementRareData::addVisibilityAdjustment):
(WebCore::ElementRareData::isVisibilityAdjustmentRoot const): Deleted.
(WebCore::ElementRareData::setIsVisibilityAdjustmentRoot): Deleted.

See above for more details — replace the single `isVisibilityAdjustmentRoot` flag with a new
`OptionSet`.

* Source/WebCore/dom/VisibilityAdjustment.h: Copied from Source/WebCore/page/ElementTargetingTypes.h.
* Source/WebCore/page/ElementTargetingController.cpp:
(WebCore::selectorsForTarget):
(WebCore::targetedElementInfo):

Teach this to emit `::before` and `::after` selectors for pseudo elements.

(WebCore::isTargetCandidate):

Teach this helper function to always consider rendered `::before` and `::after` pseudo elements to
be candidates for target selection, since adjusting visibility for these elements is unlikely to
adversely affect &quot;main content&quot; on the page.

(WebCore::elementToAdjust):
(WebCore::adjustmentToApply):
(WebCore::adjustVisibilityIfNeeded):

Split some of this functionality into multiple helper methods, to make it easier to determine how
an element should be marked for visibility adjustment. In particular, for before and after pseudo
elements, we mark the pseudo host element with `VisibilityAdjustment::{Before|After}Pseudo`, and for
everything else, we mark the element subtree (i.e. the same as what we currently do).

(WebCore::ElementTargetingController::adjustVisibility):
(WebCore::ElementTargetingController::adjustVisibilityInRepeatedlyTargetedRegions):
* Source/WebCore/page/ElementTargetingTypes.h:
* Source/WebCore/style/StyleAdjuster.cpp:
(WebCore::Style::Adjuster::adjust const):
(WebCore::Style::Adjuster::adjustVisibilityForPseudoElement):

Honor `VisibilityAdjustment::{BeforePseudo|AfterPseudo}` on the host element, by marking the
corresponding pseudo element style with visibility adjustment.

* Source/WebCore/style/StyleAdjuster.h:
* Source/WebCore/style/StyleResolver.cpp:
(WebCore::Style::Resolver::styleForPseudoElement):
* Source/WebCore/style/StyleTreeResolver.cpp:
(WebCore::Style::TreeResolver::resolveStartingStyle const):
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/UIProcess/API/APITargetedElementInfo.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKTargetedElementInfo.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKTargetedElementInfo.mm:
(-[_WKTargetedElementInfo isPseudoElement]):

Add a new SPI property to surface whether or not the targeted element is a pseudo-element to the
client.

Canonical link: <a href="https://commits.webkit.org/276816@main">https://commits.webkit.org/276816@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/7e11d2092aa101ec7c2ba235a22e430f057f1f9d

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/45727 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/24854 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/48308 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/48396 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/41762 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/48033 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/29127 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/22250 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/37437 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/46305 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/21924 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/39447 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/18589 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/19324 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/40537 "Passed tests") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/3771 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/42028 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/40876 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/50178 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/20718 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/17228 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/44578 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/22023 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/39598 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/43437 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/10161 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/22381 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/21711 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

anttijk commented on Source/WebCore/style/StyleResolver.cpp:
```diff
@@ -511,6 +519,7 @@ std::optional<ResolvedStyle> Resolver::styleForPseudoElement(const Element& elem
+    adjustVisibilityForPseudoElementIfNeeded(element, *state.style());
 
     Adjuster adjuster(document(), *state.parentStyle(), context.parentBoxStyle, nullptr);
     adjuster.adjust(*state.style(), state.userAgentAppearanceStyle());
```
This sort of code belong in `Style::Adjuster`. Currently null element is used to signal pseudo-element but it can be signaled more explicitly instead.

whsieh replied:
Chatted with Antti about this over Slack — one way to fix this would be to move this into a static helper method on `Adjuster`, i.e. `Style::Adjuster::adjustVisibilityForPseudoElement`.

In the longer run, it would be more ideal to turn the `Element*` argument in `Style::Adjuster` into `Element&` and explicitly pass in an `std::optional<PseudoElementIdentifier>` to indicate whether the element is a pseudo (instead of checking the nullity of `Element*`, like we currently do).

In attempting this strategy, however, we stumbled into an existing issue in the `@starting-style` implementation here:

```
std::unique_ptr<RenderStyle> TreeResolver::resolveStartingStyle(const ResolvedStyle& resolvedStyle, const Styleable& styleable, const ResolutionContext& resolutionContext) const
{
…
    Adjuster adjuster(m_document, parentAfterChangeStyle, resolutionContext.parentBoxStyle, styleable.pseudoElementIdentifier ? &styleable.element : nullptr);
    adjuster.adjust(*startingStyle, nullptr);
…
```

...where the nullity of the element is backwards (pseudo element currently sends a non-null element). We can fix this when we refactor the code to pass in an optional pseudo-ID and non-null `Element`, and go with the former approach for now.

whsieh committed:
[Remote Inspection] Make it possible to target ::before/::after pseudo elements
https://bugs.webkit.org/show_bug.cgi?id=271791

Reviewed by Megan Gardner.

(...and also incorporates suggestions from Antti).

Add support for targeting `::before` and `::after` pseudo elements, for visibility adjustment. See
below for more details.

* LayoutTests/fast/element-targeting/target-pseudo-elements-expected.html: Added.
* LayoutTests/fast/element-targeting/target-pseudo-elements.html: Added.

Add a new layout test to exercise the change (targeting both `after` and `before` pseudo elements).

* Source/WebCore/Headers.cmake:

Add a new WebCore header that just contains the new `VisibilityAdjustment` enum values.

* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/dom/Element.cpp:
(WebCore::Element::visibilityAdjustment const):
(WebCore::Element::addVisibilityAdjustment):
(WebCore::Element::isVisibilityAdjustmentRoot const): Deleted.
(WebCore::Element::setIsVisibilityAdjustmentRoot): Deleted.

Refactor this code — instead of having a single boolean flag that indicates whether the element is
a visibility adjustment root, turn it into an OptionSet with up to 3 bits (subtree, `::before`
pseudo element and `::after` pseudo element).

* Source/WebCore/dom/Element.h:
* Source/WebCore/dom/ElementRareData.cpp:
* Source/WebCore/dom/ElementRareData.h:
(WebCore::ElementRareData::visibilityAdjustment const):
(WebCore::ElementRareData::addVisibilityAdjustment):
(WebCore::ElementRareData::isVisibilityAdjustmentRoot const): Deleted.
(WebCore::ElementRareData::setIsVisibilityAdjustmentRoot): Deleted.

See above for more details — replace the single `isVisibilityAdjustmentRoot` flag with a new
`OptionSet`.

* Source/WebCore/dom/VisibilityAdjustment.h: Copied from Source/WebCore/page/ElementTargetingTypes.h.
* Source/WebCore/page/ElementTargetingController.cpp:
(WebCore::selectorsForTarget):
(WebCore::targetedElementInfo):

Teach this to emit `::before` and `::after` selectors for pseudo elements.

(WebCore::isTargetCandidate):

Teach this helper function to always consider rendered `::before` and `::after` pseudo elements to
be candidates for target selection, since adjusting visibility for these elements is unlikely to
adversely affect "main content" on the page.

(WebCore::elementToAdjust):
(WebCore::adjustmentToApply):
(WebCore::adjustVisibilityIfNeeded):

Split some of this functionality into multiple helper methods, to make it easier to determine how
an element should be marked for visibility adjustment. In particular, for before and after pseudo
elements, we mark the pseudo host element with `VisibilityAdjustment::{Before|After}Pseudo`, and for
everything else, we mark the element subtree (i.e. the same as what we currently do).

(WebCore::ElementTargetingController::adjustVisibility):
(WebCore::ElementTargetingController::adjustVisibilityInRepeatedlyTargetedRegions):
* Source/WebCore/page/ElementTargetingTypes.h:
* Source/WebCore/style/StyleAdjuster.cpp:
(WebCore::Style::Adjuster::adjust const):
(WebCore::Style::Adjuster::adjustVisibilityForPseudoElement):

Honor `VisibilityAdjustment::{BeforePseudo|AfterPseudo}` on the host element, by marking the
corresponding pseudo element style with visibility adjustment.

* Source/WebCore/style/StyleAdjuster.h:
* Source/WebCore/style/StyleResolver.cpp:
(WebCore::Style::Resolver::styleForPseudoElement):
* Source/WebCore/style/StyleTreeResolver.cpp:
(WebCore::Style::TreeResolver::resolveStartingStyle const):
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/UIProcess/API/APITargetedElementInfo.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKTargetedElementInfo.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKTargetedElementInfo.mm:
(-[_WKTargetedElementInfo isPseudoElement]):

Add a new SPI property to surface whether or not the targeted element is a pseudo-element to the
client.

Canonical link: https://commits.webkit.org/276816@main
